kiwi/lib/search.py

```diff
@@ -0,0 +1,655 @@
+        if output_dir.exists():
+            backup_directory = output_dir.with_name(
+                f'{output_dir.name}_backup_{datetime.now().isoformat()}'
+            )
+            logger.warning(
+                f'Folder {output_dir} already exists; moving it to {backup_directory}'
+            )
+            output_dir.rename(backup_directory)
+        output_dir.mkdir(parents=True)
+
+    logger.info(f'Initializing new search folder at: {output_dir}')
+
+    configure_logging(output_dir=output_dir, verbose=debug, quiet=quiet)
+    configure_seed(seed)
+
+    return output_dir
+
+
+def run(config: Configuration):
+    try:
+        import optuna
+    except ImportError as e:
+        raise ImportError(
+            f'{e}. Install the search dependencies with '
+            '``pip install -U openkiwi[search]``, or with '
+            '``poetry install -E search`` when setting '
+            'up for local development'
+        )
+
+    output_dir = setup_run(config.directory, config.seed)
+    logger.info(f'Saving all Optuna results to: {output_dir}')
+
+    if isinstance(config.base_config, Path):
+        # FIXME: this is not neat (but it does work)
+        #   We need this in order to support the `defaults` field in the train config,
+        #   like the
+        #     defaults:
+        #        - data: wmt19.qe.en_de
+        #   in the config/xlmroberta. It's Hydra that takes care of that inside
+        #   `kiwi.cli.arguments_to_configuration`. And because Hydra has already
+        #   been configured globally in the Kiwi cli, we need to clear it.
+        hydra._internal.hydra.GlobalHydra().clear()
+        base_dict = kiwi.cli.arguments_to_configuration(
+            {'CONFIG_FILE': config.base_config}
+        )
+        # These arguments are not in the train configuration because they are
+        #   added by `kiwi.cli.arguments_to_configuration` (and we remove them
+        #   so Pydantic won't throw an error.)
+        del base_dict['verbose'], base_dict['quiet']
+        base_config = train.Configuration(**base_dict)
+    else:
+        base_config = config.base_config
+    base_config_dict = base_config.dict()
+
+    # Perform some checks of the training config
+    use_mlflow = True if base_config_dict['run'].get('use_mlflow') else False
+    if not use_mlflow:
+        logger.warning(
+            'Using MLflow is recommend; set `run.use_mlflow=true` in the base config'
+        )
+    # The main metric should be explicitly set
+    if base_config_dict['trainer']['main_metric'] is None:
+        raise ValueError(
+            'The metric should be explicitly set in `trainer.main_metric` '
+            'in the training config (`base_config`).'
+        )
+    # The main metric(s) should be inside a list
+    if not isinstance(base_config_dict['trainer']['main_metric'], list):
+        base_config_dict['trainer']['main_metric'] = [
+            base_config_dict['trainer']['main_metric']
+        ]
```

Comment:
Why? This should be enforced by Pydantic unless there's a very specific reason to do this here

Reply:
but if I recall correctly you even use a single main metric on your objective class above